Content.Shared/Chemistry/Reaction/SharedChemicalReactionSystem.cs

```diff
@@ -132,15 +175,20 @@ protected virtual void OnReaction(Solution solution, ReactionPrototype reaction,
         /// </summary>
         private Solution ProcessReactions(Solution solution, EntityUid ownerUid)
         {
-            //TODO: make a hashmap at startup and then look up reagents in the contents for a reaction
             var overallProducts = new Solution();
-            foreach (var reaction in _reactions)
+            foreach(var reactant in solution.Contents)
             {
-                if (CanReact(solution, reaction, out var unitReactions))
+                if (!_reactions.TryGetValue(reactant.ReagentId, out var reactions))
+                    continue;
+
+                foreach(var reaction in reactions)
                 {
+                    if (!CanReact(solution, reaction, out var unitReactions))
+                        continue;
+
                     var reactionProducts = PerformReaction(solution, ownerUid, reaction, unitReactions);
                     overallProducts.AddSolution(reactionProducts);
-                    break;
+                    return overallProducts;
```

Comment:
You didn't do make it like this, but it's currently doing an unnecessary `AddSolution()`, where it is just adding `reactionProducts` to an empty solution and then returning that, when it could just return the resultts of `PerformReaction()`

Reply:
Done.